### PR TITLE
Generate: assisted generation with sample (take 2)

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

I was writing the blog post about assisted generation and realized that there is a much better way to solve the `do_sample=True` case 👀 Apologies for the repeated review request, but I believe this is a significant upgrade.

In a nutshell, the existing `temperature` argument provides a natural control mechanism for assisted generation with `do_sample=True`, when controlling how flat the distribution is at the sampling step. Lower temperature = high probability tokens become more likely to be sampled = more predictable = more likely to match the candidate tokens from the assistant model = assisted generation works faster.

Compared to the [other PR](https://github.com/huggingface/transformers/pull/22862), which was closed, this approach has the following pros and cons:
- Pros:
  - No new argument;
  - Behaves exactly like `.sample`, which users already understand well. No new heuristics;
  - As fast as the other method for similar randomness levels (see numbers below).
- Cons:
  - Internally, more than one sampling step will occur per output token. If we set a seed, the output will be different than `.sample`'s for the same seed. Not a deal breaker per se, but it means subtle bugs may be tough to catch.

## Performance numbers

  I've run the [benchmark](https://github.com/gante/huggingface-demos/tree/main/experiments/faster_generation) I've been running for assisted generation, but now for several `temperature` values. The values below are for `facebook/opt-6.7b` as the main model, `facebook/opt-125m` as the assistant model, running `.generate` starting from inputs taken from the C4 test set (i.e. quite random, the dataset I tested where assisted generation struggles the most), on a RTX3090. Note that most LLMs nowadays use temperature between 0.7 and 0.9.

TL;DR -- it's slower than greedy assisted generation, as it is expected, but it will still secure solid speedups e.g. with INT8.
  
Screenshot 2023-04-23 at 14 58 10

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Looks much cleaner, thanks!

**Review by amyeroberts — COMMENTED** [review]
The code is definitely simpler with this update 👌 

However, I think the behaviour of `temperature` here can be a bit confusing. It would be good to explain in the documentation a bit more what the temperature argument is doing and how it affects the assisted sampling.

**Review by amyeroberts — APPROVED** [review]
Sorry - I meant to hit the approve button :D

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1517,6 +1517,46 @@ def test_assisted_greedy_search_matches_greedy_search(self):
...
+                max_length=max_length,
+                num_beams=1,
+                do_sample=True,
+                assistant_model=model,  # triggers assisted decoding
+                output_scores=True,
+                output_hidden_states=True,
+                output_attentions=True,
+                return_dict_in_generate=True,
+            )
+
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1557]: The test in the other version was making a 1:1 comparison with `sample`, not with `greedy_search` :D The new structure can't make that comparison, sadly

`temperature=0` is also not an option, it causes `-inf` logits. Very low temperature values would make the test run, but it would be very flaky. Hence the decision not to compare against existing decoding methods
